app/src/main/java/emu/skyline/applet/swkbd/SoftwareKeyboardDialog.kt

```diff
@@ -0,0 +1,123 @@
+
+class SoftwareKeyboardDialog : DialogFragment() {
+    private val config by lazy { requireArguments().getParcelable<SoftwareKeyboardConfig>("config")!! }
+    private val initialText by lazy { requireArguments().getString("initialText")!! }
+    private var stopped = false
+
+    companion object {
+        /**
+         * @param config Holds the [SoftwareKeyboardConfig] that will be used to create the keyboard  between instances of this dialog
+         * @param initialText Holds the text that was set by the guest when the dialog was created
+         */
+        fun newInstance(config : SoftwareKeyboardConfig, initialText : String) : SoftwareKeyboardDialog {
+            val args = Bundle()
+            args.putParcelable("config", config)
+            args.putString("initialText", initialText)
+            val fragment = SoftwareKeyboardDialog()
+            fragment.arguments = args
+            return fragment
+        }
+
+        const val validationConfirm = 2
+        const val validationError = 1
+    }
+
+    private lateinit var binding : KeyboardDialogBinding
+
+    private var cancelled : Boolean = false
+    private var futureResult : FutureTask<SoftwareKeyboardResult> = FutureTask<SoftwareKeyboardResult> { return@FutureTask SoftwareKeyboardResult(cancelled, binding.textInput.text.toString()) }
+
+
+    override fun onCreateView(inflater : LayoutInflater, container : ViewGroup?, savedInstanceState : Bundle?) = if (savedInstanceState?.getBoolean("stopped") != true) KeyboardDialogBinding.inflate(inflater).also { binding = it }.root else null
+
+    @SuppressLint("SetTextI18n")
+    override fun onViewCreated(view : View, savedInstanceState : Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        if (config.inputFormMode == InputFormMode.OneLine) {
+            binding.header.text = stringFromChars(config.headerText)
+            binding.header.visibility = View.VISIBLE
+            binding.sub.text = stringFromChars(config.subText)
+            binding.sub.visibility = View.VISIBLE
+        }
+        if (config.keyboardMode == KeyboardMode.Numeric) {
+            binding.textInput.inputType = if (config.passwordMode == PasswordMode.Hide) InputType.TYPE_NUMBER_VARIATION_PASSWORD else InputType.TYPE_CLASS_NUMBER
+        } else {
+            binding.textInput.inputType = if (config.passwordMode == PasswordMode.Hide) InputType.TYPE_TEXT_VARIATION_PASSWORD else InputType.TYPE_CLASS_TEXT
+            if (config.invalidCharsFlags.outsideOfDownloadCode)
+                binding.textInput.inputType = binding.textInput.inputType or InputType.TYPE_TEXT_FLAG_CAP_CHARACTERS
+            if (config.isUseNewLine)
+                binding.textInput.inputType = binding.textInput.inputType or InputType.TYPE_TEXT_FLAG_MULTI_LINE
+            else
+                binding.textInput.inputType = binding.textInput.inputType and InputType.TYPE_TEXT_FLAG_MULTI_LINE.inv()
+        }
```

Comment:
Newlines around this would be appreciated, they'd semantically separate these blocks of code.